sys/firmware/firmware.c

```diff
@@ -0,0 +1,207 @@
+#define LOG_INFO(...) printf(__VA_ARGS__)
+#endif
+
+#include "firmware.h"
+#include "checksum/fletcher32.h"
+#include "tweetnacl.h"
+
+void firmware_metadata_print(firmware_metadata_t *metadata)
+{
+    printf("Firmware magic_number: 0x%08x\n", (unsigned)metadata->magic_number);
+    printf("Firmware Size: %" PRIu32 "\n", metadata->size);
+    printf("Firmware APPID: %#x\n", (unsigned)metadata->appid);
+    printf("Firmware Version: %#x\n", (unsigned)metadata->version);
+    printf("Firmware start address: 0x%08x\n", (unsigned)metadata->start_addr);
+    printf("Firmware HASH: ");
+    for (unsigned long i = 0; i < sizeof(metadata->hash); i++) {
+        printf("%02x ", metadata->hash[i]);
+    }
+    printf("\n");
+    printf("Firmware chksum: 0x%08x\n", (unsigned)metadata->chksum);
+    printf("Firmware signature: ");
+    for (unsigned long i = 0; i < sizeof(metadata->sig); i++) {
+        printf("%02x ", metadata->sig[i]);
+    }
+    printf("\n");
+}
+
+int firmware_validate_metadata_checksum(firmware_metadata_t *metadata)
+{
+    if (memcmp(metadata, "RIOT", 4)) {
+        LOG_INFO("%s: metadata magic number invalid\n", __func__);
+        return -1;
+    }
+
+    int res = firmware_metadata_checksum(metadata) == metadata->chksum ? 0 : -1;
+    if (res) {
+        LOG_INFO("%s: metadata checksum invalid\n", __func__);
+    }
+
+    return res;
+}
+
+int firmware_validate_metadata_signature(firmware_metadata_t *metadata, const unsigned char *pk)
+{
+    if (firmware_validate_metadata_checksum(metadata)) {
+        return -1;
+    }
+
+    unsigned char sm[FIRMWARE_SIGN_BYTES + crypto_sign_BYTES];
+    memcpy(sm, ((unsigned char *)metadata) + FIRMWARE_SIGN_BYTES, crypto_sign_BYTES);
+    memcpy(sm + crypto_sign_BYTES, metadata, FIRMWARE_SIGN_BYTES);
+
+    unsigned char m[FIRMWARE_SIGN_BYTES + crypto_sign_BYTES];
+    unsigned long long mlen;
+    int res = crypto_sign_open(m, &mlen, sm, FIRMWARE_SIGN_BYTES + crypto_sign_BYTES, pk);
+    if (res) {
+        LOG_INFO("%s: metadata signature invalid\n", __func__);
+    }
+    return res;
+}
+
+uint32_t firmware_metadata_checksum(firmware_metadata_t *metadata)
+{
+    return fletcher32((uint16_t *)metadata, FIRMWARE_CHECKSUM_LEN / 2);
+}
+
+int firmware_sign_metadata(firmware_metadata_t *metadata, unsigned char *sk)
+{
+    unsigned char sm[FIRMWARE_SIGN_BYTES + crypto_sign_BYTES];
+    unsigned long long smlen;
+
+    crypto_sign(sm, &smlen, (unsigned char *)metadata, FIRMWARE_SIGN_BYTES, sk);
+    memcpy(metadata->sig, sm, crypto_sign_BYTES);
+    return 0;
+}
+
+#ifdef RIOT_VERSION
+#ifndef BOARD_NATIVE
+static const unsigned _firmware_slot_start[] = {
+    CPU_FLASH_BASE,
+    CPU_FLASH_BASE + SLOT0_SIZE,
+    CPU_FLASH_BASE + SLOT0_SIZE + SLOT1_SIZE
+};
+
+const unsigned firmware_num_slots = sizeof(_firmware_slot_start) / sizeof(unsigned);
+
+void firmware_jump_to_image(firmware_metadata_t *metadata)
+{
+    uint32_t addr = (unsigned)metadata + (unsigned)sizeof(firmware_metadata_t);
```

Comment:
Why depend here on the size of the `firmware_metadata_t` (and below) and not on `FIRMWARE_METADATA_SIZE`? 
Then you could do away with the whole padding in the `firmware_metadata_t` struct and only have to ensure that it is not larger than `FIRMWARE_METADATA_SIZE`

Reply:
See kaspar030/RIOT#34